### DISCUSSION TITLE
Mapping custom collections

### DISCUSSION THREAD

**Discussion by Plasmadog** [LuckyPennySoftware/AutoMapper#3671] [NA]
I know that standard collection types are handled out-of-the-box, but it's not clear what options exist for mapping custom collection types. For example, if I have a class such as:
```
public class ImmutableListWithValueSemantics<T> : IImmutableList<T>, IEquatable<IImmutableList<T>>
{
     public ImmutableListWithValueSemantics(IImmutableList<T> list) => _list = list;
     public ImmutableListWithValueSemantics(IEnumerable<T> list) => _list = list.ToImmutableList();
     
     // etc...
}
```
What would I have to do to be able to map this type from IEnumerable<T>?

**Comment by lbargaoanu** [comment]
https://docs.automapper.org/en/latest/Custom-type-converters.html

**Reply by Plasmadog** [reply]
I saw that, but how would you use a type converter for an open generic? I guess you could do it as <object,object> and determine the correct type within the converter, but there doesn't seem to be any way for the converter to know the intended destination type in that case. The ResolutionContext doesn't appear to carry that information. The only way I can think of would be to inspect the generic type of the source and just assume that the destination should have the same type, but that seems a bit smelly.

On a side note, the docs don't really give any indication of what that ResolutionContext is meant to be useful for. No examples are shown anywhere, and the properties it exposes don't have any obvious utility.

**Reply by lbargaoanu** [reply]
https://docs.automapper.org/en/latest/Open-Generics.html

**Reply by Plasmadog** [reply]
I had seen that too, but didn't really understand it. I took another look though, and I've sort of got it working, but it's a bit of a mystery how.
I created a mapping that, based on that docs page, looks exactly like this:
```
            CreateMap(typeof(IEnumerable<>), typeof(ImmutableListWithValueSemantics<>))
                .ConvertUsing(typeof(Converter<,>));

```
Note that I used the Converter type exactly as the docs show; this is not a custom converter class that I have written. Using this mapping, when I map an instance of ImmutableListWithValueSemantics<Foo> to IEnumerable<Foo> (or List<Foo>, or Foo[]), it works perfectly. I'm not really sure how that works, but it does somehow produce a List containing all the right items. Even weirder though, is what happens when I change the mapping to this:
```
            CreateMap(typeof(IEnumerable<>), typeof(ImmutableListWithValueSemantics<>))
                .ConvertUsing(typeof(MyCustomConverter<,>));
```
where MyCustomConverter looks like this:
```
    class MyCustomConverter<TSrc,TDest> : ITypeConverter<TSrc, TDest>
    {
        public TDest Convert(TSrc source, TDest destination, ResolutionContext context)
        {
            throw new NotImplementedException();
        }
    }
```
Note the NotImplementedException in there. Now when I try mapping to IEnumerable<Foo>, it still works. No exceptions are thrown. The Convert method is never called.
At this point I wondered if the ConvertUsing was needed at all, but when I took it out it stopped working.

This is sorcery.

So, I'm happy to have a solution I can start using, but I'm no closer to having an answer to my original question. I suspect this is only working because one of my constructors takes an IEnumerable, but if that were not the case and I needed to take more control over instantiation, I would have no idea how to do it.

**Reply by lbargaoanu** [reply]
I think this is better suited for StackOverflow.
